dask/array/reductions.py

```diff
@@ -12,14 +12,34 @@
 from .core import _concatenate2, Array, atop, lol_tuples, handle_out
 from .ufunc import sqrt
 from .wrap import zeros, ones
-from .numpy_compat import divide
+from .numpy_compat import ma_divide, divide as np_divide
 from ..compatibility import getargspec, builtins
 from ..base import tokenize
 from ..context import _globals
-from ..utils import ignoring, funcname
+from ..utils import ignoring, funcname, Dispatch
 from .. import sharedict
 
 
+# Generic functions to support chunks of different types
+empty_lookup = Dispatch('empty')
+empty_lookup.register((object, np.ndarray), np.empty)
+empty_lookup.register(np.ma.masked_array, np.ma.empty)
+divide_lookup = Dispatch('divide')
+divide_lookup.register((object, np.ndarray), np_divide)
+divide_lookup.register(np.ma.masked_array, ma_divide)
+
+
+def divide(a, b, dtype=None):
+    key = lambda x: getattr(x, '__array_priority__', float('-inf'))
+    f = divide_lookup.dispatch(type(builtins.max(a, b, key=key)))
```

Comment:
We need to use `np.ma.divide` for masked arrays, and our patched version of `np.divide` for non-masked arrays. If we use `np.divide` for masked arrays, then the function applies to all elements in `.data`, which may result in `RuntimeWarning`s due to divide-by-zero. Note that this is only true for the functional operators, using the operators instead dispatches properly (e.g. `divide` vs `/`).

This is related to my question about implementing all the masked ufuncs. Normal ufuncs work and return instances of `np.ma.masked_array`, but apply to all elements in `.data` instead of only the non-masked ones. Note that even if we did implement masked ufuncs, we'd still need to do this dispatch here in the reductions code.